cms/djangoapps/contentstore/courseware_index.py

```diff
@@ -178,8 +178,10 @@ def prepare_item_index(item, skip_index=False, groups_usage_info=None):
             Returns:
             item_content_groups - content groups assigned to indexed item
             """
-            is_indexable = hasattr(item, "index_dictionary")
-            item_index_dictionary = item.index_dictionary() if is_indexable else None
+            try:
+                item_index_dictionary = item.index_dictionary()
+            except AttributeError:
```

Comment:
Alternatively, we could remove the exception handling overhead like:

```py
item_index_dictionary = getattr(item, 'index_dictionary', None)
if item_index_dictionary:
    item_index_dictionary = item_index_dictionary()
```

If we _really_ wanted, we could also check if it's a `callable` object ;)
```py
if callable(item_index_dictionary):
    item_index_dictionary = item_index_dictionary()

Reply:
```python
getattr(item, 'index_dictionary', None)
```
This would still be difficult to test because there isn't an easy way to raise `AttributeError` now that the base class has a default method. That's why the comment says that we have to _pretend_ that the method doesn't exist, by raising the exception after calling the method, rather than retrieving the method off the class.

I know we could test for `callable`, but I think the general python pattern is to assume that your object behaves as you expect, and catch exceptional cases using try/except.

Reply:
I think my biggest gripe with this is that we've now had to add special case handling to our application code to handle a contrived unittest case; that fails my smell-test.

If we need a comment to explain how this can't/won't happen outside of tests, the test is of dubious value to me.

Per the below, if we update `NonIndexableXBlock.index_dictionary` to `return None` instead of raising an exception, then this all becomes moot and this can just be:

```py
item_index_dictionary = item.index_dictionary()
```